### PR TITLE
JSON Web Token: support `nbf` and `exp` claims

### PR DESCRIPTION
**Pull request by rsto**
This updates the JSON Web Token code to

- validate the `nbf` claim
- validate the `exp` claim
- ignore unknown claims, as stated in [Section 4 of RFC7519](https://www.rfc-editor.org/rfc/rfc7519#section-4)

The initial code was contributed by @bamthomas in https://github.com/cyrusimap/cyrus-imapd/pull/4515. Thanks! This PR adds more strict validation invalid JSON claims and makes the cunit test code more maintainable.

### CODE REVIEW

File: imap/http_jwt.c [hunk 1]
```diff
@@ -503,32 +504,60 @@ static int validate_payload(struct jwt *jwt, char *out, size_t outlen)
...
+
+    if (has_err)
+        goto done;
+
+    // Validate claims
+
+    if (jnbf) {
+        time_t nbf = json_integer_value(jnbf);
+        if (!(now >= nbf)) {
+            xsyslog(LOG_ERR, "Out-of-range \"nbf\" claim", NULL);
```
Comment (ksmurchison) [inline imap/http_jwt.c:536]: Nit: Why not (now < nbf) ?

File: imap/http_jwt.c [hunk 2]
```diff
@@ -503,32 +504,60 @@ static int validate_payload(struct jwt *jwt, char *out, size_t outlen)
...
             goto done;
         }
+    }
 
-        if (max_age) {
+    if (jexp) {
+        time_t exp = json_integer_value(jexp);
+        if (!(now < exp)) {
+            xsyslog(LOG_ERR, "Out-of-range \"exp\" claim", NULL);
```
Comment (ksmurchison) [inline imap/http_jwt.c:544]: Nit: Why not (now >= exp) ?

**Review by ksmurchison — APPROVED** [review]
Other than my 2 comments LGTM

File: imap/http_jwt.c (hunk 1 above)
Comment (rsto) [reply imap/http_jwt.c:536]: @bamthomas says [here](https://github.com/cyrusimap/cyrus-imapd/pull/4515#issuecomment-1573429990):

> I used negative checks !(now < exp) and not now >= exp to be [similar to the spec wording](https://www.rfc-editor.org/rfc/rfc7519#section-4.1.5)

I guess I would have done this differently, but I'm fine with the current conditionals, too.

**Comment by elliefm** [comment]
FYI you need to use the "backport-to..." label if you want me to backport it, I don't look for plain version labels on PRs, those are for describing which versions are affected by a bug.

Is this a bug fix or a new feature?  It looks like a feature, and 3.8 is stable now and doesn't get new features, only bug fixes.

**Comment by rsto** [comment]
@elliefm thanks, did not know to use the "backport-to" label. It's a feature, not a bug. So no backport then.

**Pull request merged**